SI-8924 No outer ref in LinearSeq iterator

som-snytt opened the pull request:
The iterator for `LinearSeqLike` holds an outer pointer via
the call to `newBuilder`, which is used to empty the underlying
sequence on `toList`.

A better strategy is to use `these take 0` to create the empty
`Seq`.  We can rely on `take 0` to mean the empty `Repr`.

The unit test confirms that a weak reference to the sequence
is collectable after `seq.iterator.toList`.

som-snytt committed:
SI-8924 No outer ref in LinearSeq iterator

The iterator for `LinearSeqLike` holds an outer pointer via
the call to `newBuilder`, which is used to empty the underlying
sequence on `toList`.

A better strategy is to use `these take 0` to create the empty
`Seq`.  We can rely on `take 0` to mean the empty `Repr`.

The unit test confirms that a weak reference to the sequence
is collectable after `seq.iterator.toList`.

maxcom commented on test/junit/scala/collection/LinearSeqLikeTest.scala:
```diff
@@ -0,0 +1,85 @@
+      } finally alloc.done = true
+    assertEquals(ref, got)
+    assertNull(ref.get)
+    //Console println s"Done after ${alloc.count}"
```
I think you need to call `t.join()` to ensure that memory pressure thread is stopped before start of next unit test.

som-snytt replied:
That's an interesting suggestion.  It's harmless, but it will also have to guard against interruption; though if the allocator gets in one more allocation and it fails, even though the test has already passed, is that itself a failure?  Rhetorical question.  I'm not sure there is a notion of "next test", since usually you expect tests in parallel. But it's about ensuring clean-up.

maxcom commented on test/junit/scala/collection/LinearSeqLikeTest.scala:
```diff
@@ -0,0 +1,85 @@
+    assertEquals(two, rest.head)
+
+    // apply gentle pressure
+    class Alloc(val parent: Thread) extends Runnable {
```
This class looks dangerous, since it can't be run in parallel with other tests and also will generate heap dump if anyone run test suite with `-XX:+HeapDumpOnOutOfMemoryError` flag. I think you could allocate a large number of larger test samples (like I did in my test) and then do `System.gc() / refq.remove()` loop that awaits for garbage collection of any first sample.

som-snytt replied:
Thanks for your feedback!  The success case completes long before out of memory.  It's true that if it fails, it is not a good JVM citizen. An alternative would be for the queue.remove loop to check for heap growth, since by then you can assume failure and not wait to actually OOM.

som-snytt committed:
SI-8924 Improve test safety

Make the test less dangerous: no need to hold allocations;
stop allocating if the heap is grown, as it probably doesn't
even matter why it grew, for purposes of the test.

som-snytt commented:
This is on the barock side.  Another standalone test just allocates until an arbitrary ref is cleared; after that, your refs under test better be cleared, too.

https://github.com/scala/scala/pull/4148/files

maxcom commented:
I updated test in my #4061. Please review

Ichoran commented:
Well, we have two patches that fix the same thing, but with different tests.  Both look fine...not sure what to recommend here.  Both tests should catch the problem also.

som-snytt commented:
@maxcom @Ichoran I was curious to find a good test strategy. 

I've used this method to test  MutableList.iterator, which used to hold onto the extended tail and keep it from garbage collection.

https://github.com/scala/scala/pull/4133/files

So the machinery is refactored into `AssertUtils.assertCollectable`.

I wanted to avoid assumptions about the order in which references are cleared, and also avoid issues of indeterminacy (how long should I wait for gc to do its thing?).  Especially since the test environment can be very slow.  For instance, the thread that starts the REPL got a huge timeout because initing the compiler began to fail tests.

For this test, the timeout on queue.remove is the timeout, which is easy to reason about.  I left it at one minute (changed to one second for local testing, of course).

The other feature is to stop allocating if the heap grows.  Originally, I wanted to just fail in that case, but to accommodate running tests in parallel (or maybe even the test framework is doing allocations), now it simply stops allocating and the test just waits in the queue.remove.

If this works robustly, or can be made to work robustly, it's very handy for unit tests.

som-snytt commented:
The test can be refactored once the assert facility gets merged.
